File: apps/ios/Tapling/Environment/Configs/TaplingConfig.swift
//
//  TaplingConfig.swift
//  Tapling
//
//  Created by Benno on 02.12.25.
//

import Foundation

struct TaplingConfig {
    static let shared = TaplingConfig()

    /// Base size of the Tapling asset (width and height, since it's square).
    let baseSize: CGFloat = 248
    
    /// Base bottom offset of the body from the bottom of the image (in pixels at base size).
    let baseBodyBottomOffset: CGFloat = 83

    enum Fur: String, CaseIterable {
        case mask = "mask"
        case white = "white"
    }

    enum Hat: String, CaseIterable {
        case banana = "banana"
        case lilDuck = "lil-duck"
        case propellerHat = "propeller-hat"
    }

    enum Face: String, CaseIterable {
        case asia = "asia"
        case cute = "cute"
        case dead = "dead"
        case harryPotter = "harry-potter"
        case pilot = "pilot"
        case pixelCool = "pixel-cool"
    }

    enum Hand: String, CaseIterable {
        case down = "down"
        case up = "up"
    }

    /// Get asset name for fur base layer.
    func furBaseName(fur: Fur) -> String {
        "fur_\(fur.rawValue)_base"
    }

    /// Get asset name for fur left hand layer.
    func furLeftHandName(fur: Fur, hand: Hand) -> String {
        "fur_\(fur.rawValue)_left-\(hand.rawValue)"
    }

    /// Get asset name for fur right hand layer.
    func furRightHandName(fur: Fur, hand: Hand) -> String {
        "fur_\(fur.rawValue)_right-\(hand.rawValue)"
    }

    /// Get asset name for face layer.
    func faceName(face: Face) -> String {
        "face_\(face.rawValue)"
    }

    /// Get asset name for hat layer.
    func hatName(hat: Hat) -> String {
        "hat_\(hat.rawValue)"
    }

    private init() {}
}

/// Tapling character configuration.
struct Tapling {
    let fur: TaplingConfig.Fur
    let hat: TaplingConfig.Hat?
    let face: TaplingConfig.Face
    let leftHand: TaplingConfig.Hand
    let rightHand: TaplingConfig.Hand
}

File: apps/web/src/environment/configs/tapling.config.ts
export const taplingConfig = {
	size: 150,
	parts: {
		fur: {
			available: ['white'] satisfies TTaplingFur[],
			path: {
				base: (fur: TTaplingFur) => getTaplingPath('fur', `${fur}_base`),
				leftHand: (fur: TTaplingFur, hand: TTaplingHand) =>
					getTaplingPath('fur', `${fur}_left-${hand}`),
				rightHand: (fur: TTaplingFur, hand: TTaplingHand) =>
					getTaplingPath('fur', `${fur}_right-${hand}`)
			}
		},
		hat: {
			available: ['banana', 'lil-duck', 'propeller-hat', 'heart'] satisfies TTaplingHat[],
			path: (hat: TTaplingHat) => getTaplingPath('hat', hat)
		},
		face: {
			available: [
				'asia',
				'cute',
				'dead',
				'harry-potter',
				'pilot',
				'pixel-cool'
			] satisfies TTaplingFace[],
			path: (face: TTaplingFace) => getTaplingPath('face', face)
		},
		hand: {
			available: ['down', 'up'] satisfies TTaplingHand[]
		}
	}
};

function getTaplingPath(category: string, name: string): string {
	return `/tapling/${category}/${category}_${name}.svg`;
}

type TTaplingFur = 'mask' | 'white';

type TTaplingHat = 'banana' | 'lil-duck' | 'propeller-hat' | 'heart';

type TTaplingFace = 'asia' | 'cute' | 'dead' | 'harry-potter' | 'pilot' | 'pixel-cool';

type TTaplingHand = 'down' | 'up';

export interface TTapling {
	fur: TTaplingFur;
	hat?: TTaplingHat;
	face: TTaplingFace;
	leftHand: TTaplingHand;
	rightHand: TTaplingHand;
}
